File: src/leetcode/leetcode146.go
package leetcode

// 双向链表中的节点
type LRUDoubleListNode struct {
	prev, next *LRUDoubleListNode
	key, value int
}

// 双向链表
type LRUDoubleList struct {
	head, tail *LRUDoubleListNode
}

// 构造一个初始双向链表
func NewLRUDoubleList() LRUDoubleList {
	l := LRUDoubleList{}
	l.head = &LRUDoubleListNode{}
	l.tail = &LRUDoubleListNode{}
	l.head.next = l.tail
	l.tail.prev = l.head
	return l
}

// 将节点添加到链表首
func (l *LRUDoubleList) addToHead(node *LRUDoubleListNode) {
	headNext := l.head.next
	l.head.next = node
	headNext.prev = node
	node.next = headNext
	node.prev = l.head
}

// 移除节点
func (l *LRUDoubleList) remove(node *LRUDoubleListNode) {
	node.prev.next = node.next
	node.next.prev = node.prev
}

// 移除链表尾节点
func (l *LRUDoubleList) removeFromTail() *LRUDoubleListNode {
	if l.head.next == l.tail {
		return nil
	} else {
		node := l.tail.prev
		l.remove(node)
		return node
	}
}

// 缓存结构
type LRUCache struct {
	capacity int
	keymap   map[int]*LRUDoubleListNode
	cache    LRUDoubleList
}

// 构造一个缓存
func NewLRUCache(capacity int) LRUCache {
	lru := LRUCache{capacity: capacity}
	lru.keymap = make(map[int]*LRUDoubleListNode)
	lru.cache = NewLRUDoubleList()
	return lru
}

// 设某个键为最近访问
func (lru *LRUCache) setMostRecently(key int) {
	node := lru.keymap[key]
	lru.cache.remove(node)
	lru.cache.addToHead(node)
}

// 新增元素
func (lru *LRUCache) addItem(key, Value int) {
	node := &LRUDoubleListNode{key: key, value: Value}
	lru.keymap[key] = node
	lru.cache.addToHead(node)
}

// 移除元素
func (lru *LRUCache) removeItem(key int) {
	node := lru.keymap[key]
	lru.cache.remove(node)
	delete(lru.keymap, key)
}

// 移除最久未访问元素
func (lru *LRUCache) removeLeastRecently() {
	node := lru.cache.removeFromTail()
	delete(lru.keymap, node.key)
}

// 读取操作
func (lru *LRUCache) Get(key int) int {
	if item, ok := lru.keymap[key]; ok {
		lru.setMostRecently(item.key)
		return item.value
	} else {
		return -1
	}

}

// 新增操作
func (lru *LRUCache) Put(key int, value int) {
	if item, ok := lru.keymap[key]; ok {
		item.value = value
		lru.setMostRecently(item.key)
	} else {
		if len(lru.keymap) >= lru.capacity {
			lru.removeLeastRecently()
		}
		lru.addItem(key, value)
	}
}

func Constructor146(capacity int) LRUCache {
	return NewLRUCache(capacity)
}

File: src/leetcode/leetcode146.py
from typing import Dict,  Optional, Union


class DoubleListNode(object):
    '''
    存储的键值对双向链表节点
    Attributes:
        key: 缓存的键
        value: 缓存的值
        prev: 前指针
        node: 后指针
    '''

    def __init__(self, key: Optional[int] = None, value: Optional[int] = None):
        self.key: Optional[int] = key
        self.value: Optional[int] = value
        self.prev: Optional[DoubleListNode] = None
        self.next: Optional[DoubleListNode] = None


class DoubleList(object):
    '''
    存储的键值对双向链表

    Attributes:
        head: 头指针，不包含有用数据，即哑节点
        tail: 尾指针，不包含有用数据，即哑节点
    '''

    def __init__(self):
        self.head = DoubleListNode()
        self.tail = DoubleListNode()
        self.head.next = self.tail
        self.tail.prev = self.head

    def add_to_head(self, node: DoubleListNode):
        '''
        添加新节点到双向节点的头部
        Args:
            node: 新节点。这个节点应该在调用前创建完成并传入
        Returns:
            None
        Raise:
            AttributeError: 找不到头节点的后节点
        '''
        head_next = self.head.next
        if head_next is None:
            raise AttributeError("head_next is None")
        self.head.next = node
        head_next.prev = node
        node.next = head_next
        node.prev = self.head

    def remove(self, node: DoubleListNode):
        '''
        从链中删除一个节点
        Args:
            node: 要删除的节点。这个节点应该在调用前通过其他方法找到其实例并传入
        Returns:
            None
        Raise:
            AttributeError: 找不到node的前后节点
        '''
        if node.prev is None or node.next is None:
            raise AttributeError("node next or prev is None")
        node.prev.next = node.next
        node.next.prev = node.prev

    def remove_from_tail(self) -> Union[DoubleListNode, None]:
        '''
        删除链表尾的节点
        Args:
            None
        Returns:
            如果链表包含有用数据，则返回被删除的节点实例。否则返回None
        Raise:
            AttributeError: 找不到tail的前节点
        '''
        if self.head.next == self.tail:
            return None
        else:
            if self.tail.prev is None:
                raise AttributeError("tail prev is None")
            node: DoubleListNode = self.tail.prev
            self.remove(node)
            return node


class LRUCache(object):
    '''
    LRU 缓存类，
    Attributes:
        cache: 缓存链表，每个节点存储了该缓存的键值对
        capacity: 容量，超过此容量将会抛弃最久未使用的缓存节点
        keymap: 哈希表，每个键对应一个缓存节点
    '''

    def __init__(self, capacity):
        '''
        初始化，指定容量
        '''
        self.capacity = capacity
        self.keymap: Dict[int, DoubleListNode] = {}
        self.cache = DoubleList()

    def add_item(self, key: int, value: int):
        '''
        添加新缓存
        Args:
            key: 缓存的键
            value: 缓存的值
        Returns:
            None
        Raise:
            None
        '''
        node = DoubleListNode(key=key, value=value)
        self.keymap[key] = node
        self.cache.add_to_head(node)

    def remove_item(self, key: int):
        '''
        删除缓存
        Args:
            key: 缓存的键
        Returns:
            None
        Raise:
            None
        '''
        node = self.keymap[key]
        self.cache.remove(node)
        del self.keymap[key]

    def remove_least_recently(self):
        '''
        删除最久未使用缓存
        Args:
            None
        Returns:
            None
        Raise:
            None
        '''
        node = self.cache.remove_from_tail()
        if node is None or node.key is None:
            return
        del self.keymap[node.key]

    def set_most_recently(self, key: int):
        '''
        设定某个键对应的缓存为最近使用
        Args:
            key: 缓存的键
        Returns:
            None
        Raise:
            None
        '''
        node = self.keymap[key]
        self.cache.remove(node)
        self.cache.add_to_head(node)

    def get(self, key: int) -> int:
        '''
        读取缓存
        Args:
            key: 缓存的键
        Returns:
            若存在该缓存，则返回该缓存的值，否则，返回-1
        Raise:
            None
        '''
        if key in self.keymap:
            value = self.keymap[key].value
            if value is not None:
                self.set_most_recently(key)
                return value
        return -1

    def put(self, key: int, value: int):
        '''
        新增缓存
        Args:
            key: 缓存的键
            value: 缓存的值
        Returns:
            None
        Raise:
            None
        '''
        if key in self.keymap:
            self.keymap[key].value = value
            self.set_most_recently(key)
        else:
            if len(self.keymap) >= self.capacity:
                self.remove_least_recently()
            self.add_item(key, value)


if __name__ == '__main__':
    lru = LRUCache(2)
    lru.put(1, 1)
    lru.put(2, 2)
    lru.get(1)
    lru.put(3, 3)
    lru.get(2)
    lru.put(4, 4)
    lru.get(1)
    lru.get(3)
    lru.get(4)
